### DISCUSSION TITLE
variables mapping between original and preprocessed models

### DISCUSSION THREAD

**Discussion by nnarodytska** [google/or-tools#3274] [CP-SAT questions]
Dear All,

[Problem]  MaxSAT, so all variables are Booleans.

[Or-tools solver]  CP-SAT solver (I do not use maxsat-specific flags. A preprocessor is on.)

Question: Can you please help me understand how variables of the original model are mapped to variables of a preprocessed model (new_cp_model_proto defined in line [1])? It looks like "postsolve_mapping" (defined in line [2]) stores such a mapping but I am not sure. My goal is to extract Boolean variables that are fixed during the search in the preprocessed model (these are variables fixed at  the decision level 0 in the trail of a SAT solver) that correspond to variables of the original model. My original problem is MaxSAT, so I think many fixed variables in the preprocessed model should correspond to variables of the original problem (as we do not need to encode integers domains)

[1] https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/sat/cp_model_solver.cc#L3264
[2] https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/sat/cp_model_solver.cc#L3433

Any help will be very much appreciated!

Thank you!

nina

**Comment by lperron** [comment]
Hi Nina, 

Actually, the solver first presolve the model. Then start propagating at the root node on the presolved model.
These are 2 different phases.

The mapping model contains all original variables, as well as constraints that allows recomputing all non-fixed variables from the solution of the presolved model into the original model.

the presolved mapping is just the mapping from presolved variable into original variable. This is used to inject the solution of the presolved model into the mapping model.

All in all, nothing there contains the variables fixed at the root node of the search.

**Comment by nnarodytska** [comment]
Hi Laurent,

Thank you very much for clarifications! I thought that variables of the presolved model form a subset of variables in SAT solver.
May I ask you one more clarification?

I observe that there are a lot of variables that are fixed during search in the second phase (after presolve).
These are variables that have decision level 0 in trail (`model->Get<Trail >()`).
Is there any correspondence between these "fixed bools" and Boolean variables of the original problem?
Here is a snippet of logs/prints [1 worker is used]. Initial problem has 43861 variables.
After presolve, we have 27870 variables. However, during the search, a large number of them is fixed.


Preloading model
----- the number of vars in model_proto and  new_cp_model_proto at this point are ---------->  new_cp_model_proto.variables().size()= 27870,  mapping_proto.variables().size()= 43861,  model_proto.variables().size()= 43861
[Symmetry] Graph for symmetry has 357517 nodes and 1242922 arcs.
[Symmetry] Symmetry computation done. time: 0.215181 dtime: 0.186059
#Bound   7.10s best:inf   next:[0,300]    initial_domain
Starting to load the model at 7.10s
[Probing] deterministic_time: 1.00072 (limit: 1) wall_time: 0.555627 (Aborted 13095/27870)
[Probing]  - new fixed Boolean: 6 (6/27870)
[Probing]  - new binary clause: 12775
[LinearRelaxationBeforeCliqueExpansion] #linear:0 #at_most_ones:7525
Initial num_bool: 27870
Starting sequential search at 8.39s
................
#35      9.05s best:60    next:[2,59]      fixed_bools:6473/27870
#36      9.06s best:59    next:[2,58]      fixed_bools:6683/27870
#37      9.07s best:58    next:[2,57]      fixed_bools:6683/27870
#38      9.08s best:57    next:[2,56]      fixed_bools:6683/27870
#39      9.09s best:56    next:[2,55]      fixed_bools:6683/27870
#40      9.13s best:53    next:[2,52]      fixed_bools:6836/27870
#41      9.14s best:52    next:[2,51]      fixed_bools:6836/27870


To give you some context if you are interested. I am looking at some hard instances for core-guided solvers.  I noticed that ortools finds a very good upper bound for these instances (optimal in many cases) but it cannot prove optimality.
I was wondering if I get an upper bound from or-tools and all implied variables (if any) and pass it to a core-guided maxsat solver, whether I can solve these instances.

Thank you!

nina

**Comment by lperron** [comment]
First advice, use a beefy computer and enable multiple workers.

Parameter: num_search_workers = 8 (or 16 if you have enough cores).

**Reply by nnarodytska** [reply]
Hi Laurent,

Thank you for the reply!

I did try multiple workers (10) on some of these hard instances but it did not help with proving optimality given 1 hour timeout.  I am experimenting with a new rc2/maxres hybrid solver  that pushes the lower bound faster at least in my experiments.  (Also, we cannot use multiple threads on competition  benchmarks to compare to other solvers.)  

That is why I am curious if I use ortools to get an upper bound and implied vars, would it be sufficient help for our solver on these benchmarks?  I can share a subset of maxsat (unweighted, competition) benchmarks that are hard for core-guided (but maxhs can solve them) that I am looking at if you are interested  for benchmarking.

Thanks!

nina

**Comment by lperron** [comment]
just a note, max hs is mostly give everything to CPLEX (if the problem is not too big) :-)

**Reply by nnarodytska** [reply]
Yes. In fact, many competition  benchmarks can be solved by gurobi in 5 minutes ( with 1 thread). But this subset  that I focus on is different, maxhs solves them in about 40-60 minutes, gurobi does not solve them.

**Reply by nnarodytska** [reply]
To clarify, I am talking about unweighted MaxSAT benchmarks.

**Comment by lperron** [comment]
second remark, in the minizinc challenge we run interleaved workers on a single thread.

These are the parameters we set:
```
      parameters.set_search_branching(SatParameters::AUTOMATIC_SEARCH);
      parameters.set_interleave_search(true);
      parameters.set_reduce_memory_usage_in_interleave_mode(true);
      parameters.set_num_search_workers(1);
```

**Reply by nnarodytska** [reply]
Thanks, I will have a look!  I was running 10 workers for 1 hour, I could not solve some of these  benchmarks.

**Comment by nnarodytska** [comment]
Laurent, sorry for bothering with this. Can you please help me understand if there is any relation between these "fixed bools" and Boolean variables of the original problem? Thank you!

**Comment by lperron** [comment]
fixed bool are discovered by probing during the loading of the problem.

There is a first probing round during presolve, and a second one when
starting the search.

These are variables from the presolved model. Nothing forbids the presolved
model to have additional variables w.r.t. the original model.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by nnarodytska** [reply]
Thanks a lot for the reply! 

The very last question: is there a data-structure that let me map variabels of a presolved model to variables of the original model (for those that have  such a mapping)?  Or do I need  to monitor probing and record that a variable is fixed?

If I manage to run these exps: "maxsat input" ->ortools -> [get an upper bound and implied vars]->run maxres-rc2 then I will share results. I have about 10 instances that I am focusing on for unweighted maxsat. I think we have many more weighted MaxSAT instances that are hard for core-guided as some instances have very large optimal solutions and relatively small cost per soft clause.

Thank you!

nina

**Comment by nnarodytska** [comment]
@NikolajBjorner +Nikolaj as he might be interested. It looks like probing is very useful for some unweighted maxsat instances.

**Reply by lperron** [reply]
Commenting before closing the discussion. Since, we have added more probing and objective shaving workers: probing_max_lp, probing_no_lp, objective_lb_search, and objective_shaving_search.
